csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -19,6 +19,182 @@
+  // consumer.
+  // 3. Multiple splits or merge reshapes: [x, y, z] -> [xyz]. Sharding on x and
+  // xyz. Similarly for the corresponding split reshape.
+  // 4. Independent splits or merge reshapes: [w, x, y, z] -> [wx, yz]. Sharding
+  // is on w and y. In the consumer, it is applied to wx and yz.
+  // An improvement is to support mult-levels of sharding (not a real case yet)
+  // if they are all outer splits. For example: For the reshape [h] -> [a, h/a]
+  // where the h is sharded twice: [h] -> [cp, h/cp] -> [cp, tp, h/(cp*tp)]
+
+  // A more general approach maybe to "undo" the reshape (reverse transforms
+  // from root to logical domain), followed by simplification of the consumer
+  // loop domain to move DID upwards.
+
+  TensorView* producer = view_op->in();
+  TensorView* consumer = view_op->out();
+
+  const std::unordered_map<IterDomain*, IterDomain*>& c2p =
+      PairwiseLogicalDomainMap(producer, consumer).mapConsumerToProducer();
+  const std::unordered_map<IterDomain*, IterDomain*>& p2c =
+      PairwiseLogicalDomainMap(producer, consumer).mapProducerToConsumer();
+  auto [p_logical_reshaped_ids, c_root_reshaped_ids] =
+      getReshapedIds(view_op, c2p);
+
+  auto p_loop_domain = producer->getLoopDomain();
+
+  // Track number of DID axis on reshaped ids that were propagated to the
+  // consumer. These will not be included later in TransformReplay.
+  int64_t num_reshape_shardings = 0;
+
+  for (auto idx : c10::irange(did_pos)) {
+    IterDomain* p_did = p_loop_domain.at(idx);
+    NVF_ERROR(p_did->isDeviceDim());
+
+    auto p_transforms = DependencyCheck::getAllExprsBetween(
+        {p_logical_reshaped_ids.begin(), p_logical_reshaped_ids.end()},
+        {p_loop_domain.at(idx)});
+
+    if (p_transforms.empty()) {
+      // This device axis is not on reshaped ids.
+      continue;
+    }
+
+    if (p_transforms.size() > 1) {
+      // This reshape has been transformed.
+      // This can happen, for example, when there is a consumer-to-producer
+      // propagation before this pass. We will attempt using TransformReplay
+      // to propagate the sharding.
+      // TODO: Support multi-level shardings for cp + tp.
+      continue;
+    }
+
+    NVF_ERROR(
+        p_transforms.front()->isA<Split>(),
+        "Expected a split transform producing the did axis.");
+
+    num_reshape_shardings++;
+
+    // Find the producer logical id that is sharded.
+    // We expect the outermost reshaped id to be outer-split and follow the
+    // outermost path traversing the transforms. We do not support inner or
+    // non-divisible splits.
+    auto* p_did_split = p_did->definition()->as<Split>();
+    NVF_ERROR(
+        isSplitDivisible(p_did_split->in(), p_did_split),
+        "Expected the split to be divisble. Got: ",
+        p_did_split->in(),
+        " and split factor: ",
+        p_did_split->factor());
+    NVF_ERROR(!p_did_split->innerSplit(), "Inner split is not supported.");
```

Comment:
This looks wrong. IIUC, you want to check that `p_did` is the outer of the split. It's different from checking that the split is an outer split. For example, consider inner-splitting `h` by `h/d` and parallelizing `d` on DIDx.

Reply:
> IIUC, you want to check that p_did is the outer of the split. 

A correction: I think this has to be in **addition** to your existing `!innerSplit()`. The reason is below. 

> inner-splitting `h` by `h/d` and parallelizing `d` on DIDx

IIUC, this needs to be rejected as well. Yes, `d` is the outer and is parallelized. But your current implementation won't be able to **replay** the split on another size as we would hope for. For example, replaying the split on `a` will give us `a/(h/d), h/d`, not `d, a/d`. 

That said, the current check is still problematic because it incorrectly accepts an outer-split of `h` into `[h/d, p_did=d]`